ogr/ogrsf_frmts/miramon/ogrmiramondatasource.cpp

```diff
@@ -0,0 +1,310 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE. 
+ ****************************************************************************/
+
+#include "ogrmiramon.h"
+
+/****************************************************************************/
+/*                          OGRMiraMonDataSource()                          */
+/****************************************************************************/
+OGRMiraMonDataSource::OGRMiraMonDataSource()
+    : papoLayers(nullptr), nLayers(0), pszRootName(nullptr), pszDSName(nullptr),
+      bUpdate(false)
+
+{
+    MMMap.nNumberOfLayers = 0;
+    MMMap.fMMMap = nullptr;
+}
+
+/****************************************************************************/
+/*                         ~OGRMiraMonDataSource()                          */
+/****************************************************************************/
+
+OGRMiraMonDataSource::~OGRMiraMonDataSource()
+
+{
+    for (int i = 0; i < nLayers; i++)
+        delete papoLayers[i];
+    CPLFree(papoLayers);
+    CPLFree(pszDSName);
+    CPLFree(pszRootName);
+
+    if (MMMap.fMMMap)
+        VSIFCloseL(MMMap.fMMMap);
+}
+
+/****************************************************************************/
+/*                                Open()                                    */
+/****************************************************************************/
+
+int OGRMiraMonDataSource::Open(const char *pszFilename, VSILFILE *fp,
+                               const OGRSpatialReference *poSRS, int bUpdateIn,
+                               CSLConstList papszOpenOptionsUsr)
+
+{
+    bUpdate = CPL_TO_BOOL(bUpdateIn);
+
+    OGRMiraMonLayer *poLayer = new OGRMiraMonLayer(
+        pszFilename, fp, poSRS, bUpdate, papszOpenOptionsUsr, &MMMap);
+    if (!poLayer->bValidFile)
+    {
+        delete poLayer;
+        return FALSE;
+    }
+    papoLayers = static_cast<OGRMiraMonLayer **>(CPLRealloc(
+        papoLayers,
+        (size_t)(sizeof(OGRMiraMonLayer *) * ((size_t)nLayers + (size_t)1))));
+    papoLayers[nLayers] = poLayer;
+    nLayers++;
+
+    if (pszDSName)
+    {
+        const char *pszExtension = CPLGetExtension(pszDSName);
+        if (!EQUAL(pszExtension, "pol") && !EQUAL(pszExtension, "arc") &&
+            !EQUAL(pszExtension, "pnt"))
+        {
+            strcpy(
+                MMMap.pszMapName,
+                CPLFormFilename(pszDSName, CPLGetBasename(pszDSName), "mmm"));
+            if (!MMMap.nNumberOfLayers)
+            {
+                MMMap.fMMMap = VSIFOpenL(MMMap.pszMapName, "w+");
```

Comment:
return of VSIFOpenL should be checked  against nullptr

Reply:
Done